File: pkg-r/tests/testthat/test-dep-discovery.R
# Automatic output dependency discovery (#146/#203). Python's Express-mode
# counterpart (inline harvest at page-generation time) is asserted in
# pkg-py/tests/test_page_dep_harvest.py; R must push post-flush instead, since
# the UI renders before server() runs.

fake_dep <- function(name, version = "1.0.0") {
  htmltools::htmlDependency(
    name = name,
    version = version,
    src = c(href = name),
    script = paste0(name, ".js")
  )
}

# A render function whose outputFunc attaches `dep`, standing in for e.g.
# plotly::renderPlotly() without the test dependency.
fake_render_fn <- function(dep) {
  shiny::createRenderFunction(
    function() "value",
    transform = function(value, session, name, ...) value,
    outputFunc = function(id) {
      htmltools::attachDependencies(htmltools::div(id = id), dep)
    }
  )
}

# Minimal stand-in for the slice of ShinySession that install_dep_discovery()
# touches: userData, onFlushed, getOutput, sendCustomMessage, and the private
# `.outputs` list.
fake_session <- function() {
  private <- new.env()
  private$.outputs <- list()
  flush_callbacks <- list()
  messages <- list()
  session <- new.env()
  session$userData <- new.env()
  session$.__enclos_env__ <- list(private = private)
  session$onFlushed <- function(callback, once = FALSE) {
    flush_callbacks[[length(flush_callbacks) + 1]] <<- callback
  }
  session$getOutput <- function(name) {
    attr(private$.outputs[[name]], "renderFunc", exact = TRUE)
  }
  session$sendCustomMessage <- function(type, message) {
    messages[[length(messages) + 1]] <<- list(type = type, message = message)
  }
  session$define_output <- function(name, render_fn) {
    obs <- structure(list(), renderFunc = render_fn)
    private$.outputs[[name]] <- obs
  }
  session$flush <- function() {
    for (callback in flush_callbacks) {
      callback()
    }
  }
  session$messages <- function() messages
  session
}

sent_dep_names <- function(msg) {
  vapply(msg$message, function(dep) dep$name, character(1))
}

test_that("discovery pushes new outputs' deps once per flush cycle", {
  session <- fake_session()
  expect_true(install_dep_discovery(session))

  session$define_output("plot1", fake_render_fn(fake_dep("fake-binding")))
  session$flush()
  msgs <- session$messages()
  expect_length(msgs, 1)
  expect_identical(msgs[[1]]$type, "shinyreact-deps")
  expect_identical(sent_dep_names(msgs[[1]]), "fake-binding")

  # No new outputs => no message on later flushes.
  session$flush()
  expect_length(session$messages(), 1)
})

test_that("discovery covers outputs registered after startup, without resends", {
  session <- fake_session()
  install_dep_discovery(session)

  session$define_output("a", fake_render_fn(fake_dep("binding-a")))
  session$flush()

  # A late-mounted output (e.g. module server inside an observer): its new
  # dep is pushed, the already-sent one is not.
  session$define_output("b", fake_render_fn(fake_dep("binding-a")))
  session$define_output("c", fake_render_fn(fake_dep("binding-c")))
  session$flush()

  msgs <- session$messages()
  expect_length(msgs, 2)
  expect_identical(sent_dep_names(msgs[[2]]), "binding-c")
})

test_that("dep-less outputs (renderText, reactive_output) push nothing", {
  session <- fake_session()
  install_dep_discovery(session)
  session$define_output("txt", shiny::renderText("hi"))
  session$define_output("react", reactive_output(42))
  session$flush()
  expect_length(session$messages(), 0)
})

test_that("install_dep_discovery() is idempotent per session", {
  session <- fake_session()
  expect_true(install_dep_discovery(session))
  expect_false(install_dep_discovery(session))
  session$define_output("a", fake_render_fn(fake_dep("binding-a")))
  session$flush()
  expect_length(session$messages(), 1)
})

test_that("install_dep_discovery() no-ops without a real session", {
  expect_false(install_dep_discovery(NULL))
  expect_false(install_dep_discovery(list(userData = list())))
})

test_that("input handlers install discovery", {
  session <- fake_session()
  default_input_handler(list(), session = session)
  expect_true(session$userData[[".shinyreact_dep_discovery"]])

  session2 <- fake_session()
  asis_input_handler(1, session = session2)
  expect_true(session2$userData[[".shinyreact_dep_discovery"]])
})

File: pkg-py/tests/test_dep_discovery.py
"""Core-mode automatic output dependency discovery (#220).

Mirrors R's pkg-r/tests/testthat/test-dep-discovery.R. The Express-mode
counterpart (inline harvest at page-generation time) is in
test_page_dep_harvest.py; Core-mode pages render before server() runs, so deps
must be pushed after each flush instead.
"""

from __future__ import annotations

import asyncio
from typing import Any, Callable

from htmltools import HTMLDependency, TagChild, TagList, div
from shiny.express._stub_session import ExpressStubSession
from shiny.input_handler import input_handlers
from shiny.render.renderer import Renderer
from shiny.session import session_context
from shinyreact._dep_discovery import install_dep_discovery


class _FakeSession(ExpressStubSession):
    """ExpressStubSession + the slice discovery touches: flush callbacks,
    custom messages, and a real (route-less) _process_ui."""

    def __init__(self) -> None:
        super().__init__()
        self.flush_callbacks: list[Callable[[], Any]] = []
        self.messages: list[tuple[str, Any]] = []

    def on_flushed(self, fn, once: bool = True):  # type: ignore[override]
        self.flush_callbacks.append(fn)
        return lambda: None

    def _process_ui(self, ui: TagChild):  # type: ignore[override]
        deps = TagList(ui).render()["dependencies"]
        return {"deps": [d.as_dict(lib_prefix="lib") for d in deps], "html": ""}

    async def send_custom_message(self, type: str, message: Any) -> None:
        self.messages.append((type, message))

    def flush(self) -> None:
        for fn in self.flush_callbacks:
            asyncio.run(fn())


def _fake_dep(name: str, version: str = "1.0.0") -> HTMLDependency:
    return HTMLDependency(name, version, source={"href": name}, script={"src": "x.js"})


def _define_output(
    session: _FakeSession, name: str, dep: HTMLDependency | None
) -> None:
    class render_widget(Renderer[str]):  # noqa: N801
        def auto_output_ui(self):
            return div(dep, id=self.output_id) if dep is not None else None

        async def transform(self, value: str) -> str:  # pragma: no cover
            return value

    def value_fn() -> str:  # pragma: no cover - never rendered
        return "value"

    value_fn.__name__ = name  # the output id comes from the function name
    with session_context(session):
        render_widget(value_fn)


def _sent_dep_names(message: Any) -> list[str]:
    return [dep["name"] for dep in message]


def test_discovery_pushes_new_output_deps_once_per_flush() -> None:
    session = _FakeSession()
    assert install_dep_discovery(session) is True

    _define_output(session, "plot1", _fake_dep("fake-binding"))
    session.flush()
    assert len(session.messages) == 1
    assert session.messages[0][0] == "shinyreact-deps"
    assert _sent_dep_names(session.messages[0][1]) == ["fake-binding"]

    # No new outputs => no message on later flushes.
    session.flush()
    assert len(session.messages) == 1


def test_discovery_covers_late_outputs_without_resends() -> None:
    session = _FakeSession()
    install_dep_discovery(session)

    _define_output(session, "a", _fake_dep("binding-a"))
    session.flush()

    # A late-mounted output (e.g. a module server inside an observer): its new
    # dep is pushed, the already-sent one is not.
    _define_output(session, "b", _fake_dep("binding-a"))
    _define_output(session, "c", _fake_dep("binding-c"))
    session.flush()

    assert len(session.messages) == 2
    assert _sent_dep_names(session.messages[1][1]) == ["binding-c"]


def test_dep_less_outputs_push_nothing() -> None:
    session = _FakeSession()
    install_dep_discovery(session)
    _define_output(session, "txt", None)
    session.flush()
    assert session.messages == []


def test_install_is_idempotent_per_session() -> None:
    session = _FakeSession()
    assert install_dep_discovery(session) is True
    assert install_dep_discovery(session) is False
    _define_output(session, "a", _fake_dep("binding-a"))
    session.flush()
    assert len(session.messages) == 1


def test_install_no_ops_without_a_real_session() -> None:
    assert install_dep_discovery(None) is False
    assert install_dep_discovery(object()) is False  # type: ignore[arg-type]


def test_init_handler_installs_discovery() -> None:
    session = _FakeSession()
    assert input_handlers["shinyreact.init"](1, "x", session) == 1
    assert session.flush_callbacks != []


def test_value_handlers_do_not_install_discovery() -> None:
    session = _FakeSession()
    input_handlers["shinyreact.default"]([], "x", session)
    input_handlers["shinyreact.asis"](1, "x", session)
    assert session.flush_callbacks == []
